src/EFCore.MySql/Storage/ValueComparison/Internal/JsonValueComparer.cs

```diff
@@ -0,0 +1,16 @@
+﻿using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Microsoft.EntityFrameworkCore.Storage.ValueConversion;
+
+namespace Pomelo.EntityFrameworkCore.MySql.Storage.ValueComparison.Internal
+{
+    public class JsonValueComparer<T> : ValueComparer<T>
+    {
+        public JsonValueComparer(ValueConverter valueConverter)
+            : base(
+                (left, right) => object.Equals(valueConverter.ConvertToProvider(left), valueConverter.ConvertToProvider(right)),
```

Comment:
Equals lambda is also on the hot path.

Reply:
The only general way I could think about speeding this up was to traverse the object graphs ourselves in parallel, so we can compare them node by node and possibly break early, if they diverge. This would however be significantly more complex then just letting the JSON library do the work and comparing the result.

For Newtonsoft.Json, if `T` is a DOM class, then [JTokenEqualityComparer](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Linq_JTokenEqualityComparer.htm) could be used.

Reply:
You could detect whether it implements `IEquatable<T>` or overrides `Equals` and use that.

Reply:
> You could detect whether `T` implements `IEquatable<T>` or overrides `Equals` and use that.

Good points!